### ISSUE TITLE
RuntimeError: abort(out of memory)

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#1513]
As reported by @jsp1 in https://github.com/pyodide/pyodide/issues/1331#issuecomment-823693684

> I've been using cdn.jsdelivr.net/pyodide/dev/full today, and have several times run into things like this:

![image](https://user-images.githubusercontent.com/630936/115511548-3d6d8280-a281-11eb-9d6b-a3f6761270b1.png)

> I've seen out-of-memory issues occasionally before, but usually I'd close the tab, reopen, and it would be better. These seem a bit more persistent. Also, my understanding is that there were fixes in v0.17 that were intended to address this, so... I guess it's not working for me (and may have made things worse)? I don't really have a consistent repro case, but if there's any info I can provide to help you debug, let me know.

> I'm on Firefox 87.0.
> [...]

> Yes, closing and restarting firefox does appear to fix it. I also poked around in "about:memory" to see if there was anything useful. Perhaps this is 
![image](https://user-images.githubusercontent.com/630936/115511722-74dc2f00-a281-11eb-9e56-fd3a83c6020b.png)

> Anytime I see things going wrong near 4GB, I start to get suspicious there's some 32-bit limit rearing its head...

**Comment by rth** [comment]
4GB of memory is indeed the the limit for 32bit WebAssembly VM as indicated in https://github.com/WebAssembly/spec/pull/1121

I'm not sure if all browsers already implemented the extension from 2GB to 4GB as this is fairly recent. For instance in Chrome they extended it in 2020 https://v8.dev/blog/4gb-wasm-memory

**Comment by rth** [comment]
Actually the maximum memory that can be allocated in Pyodide is still 2GB, as we compile wit that value of [`MAXIMUM_MEMORY`](https://github.com/emscripten-core/emscripten/blob/5465016fb1329f97f583e534280ddba80a876595/src/settings.js#L192) and increasing it over this limit increases code size + not all browsers support it.

**Comment by jps1** [comment]
Thanks for looking into this!  A couple questions:

- I am by no means doing anything large/complex that should be using 4GB (or even 2GB) of memory.  Instead, this generally happens after I reload the page a number of times, so seems like I'm hitting #1333.  It used to be pretty easy to workaround, as closing the tab and reopening it would fix it, but the last couple days, even that doesn't help and I have to close Firefox completely and restart it (which is much less convenient, since I'm a tab hoarder, so restarting is... not fast).  I'm not sure whether this change in required workaround is because I upgraded Firefox or pyodide (I suspect the former, though not sure).  I also confirmed that after I close the tab, "about:memory" still shows those allocations.  Do you have any suggestions for lighter-weight workarounds or ways to avoid this?  Is there anything I can do in my application to increase the likelihood of old stuff being recycled after page-reload?  Or is this just a bug in Firefox that we have to wait for them to fix? 😞 
- Python isn't exactly known for being light on memory usage, so I can certainly imagine a scenario where I _do_ want to use >4GB of memory.  If/when I get to that point, is there a way to run pyodide with wasm64 (even if I have to do my own build, etc)?

**Comment by rth** [comment]
> It used to be pretty easy to workaround, as closing the tab and reopening it would fix it, but the last couple days, even that doesn't help and I have to close Firefox completely and restart it
> ...
> Or is this just a bug in Firefox that we have to wait for them to fix?

Yes, I think we should open a Firefox bug (after checking that there isn't one open already). There isn't really much we can do about how memory is manged on tab reload on the Pyodide side.. 

>  Is there anything I can do in my application to increase the likelihood of old stuff being recycled after page-reload?

From https://github.com/pyodide/pyodide/pull/1306 it looks it's possible to force garbage collection from "about:memory". There might be also some settings in `about:config` to make the garbage collection more eager. We need to investigate. But really we shouldn't have to do this, we should report it to Firefox.

> I can certainly imagine a scenario where I do want to use >4GB of memory. If/when I get to that point, is there a way to run pyodide with wasm64

For Pyodide it would also be preferable if we could use wasm64, but for now we are still waiting for toolchain and browser support. There are some steps in that direction but hard to tell when it will happen https://github.com/emscripten-core/emscripten/issues/12087

**Comment by hoodmane** [comment]
@jps1 What I do is install a separate version of Firefox (or a Chromium) and move Pyodide to a different browser when I get problems.

**Comment by hoodmane** [comment]
> it would also be preferable if we could use wasm64

We'd have to change all the code in core that implicitly assumes 32 bit addresses. =)

**Comment by rth** [comment]
Actually I also get this error when importing numpy on Firefox 74.0.1 (released on 2020-03-10) 32bit on Windows. And we shouldn't be reaching 2GB RAM there.
